app/scripts/controllers/ens/index.js

```diff
@@ -0,0 +1,75 @@
+const ethUtil = require('ethereumjs-util')
+const ObservableStore = require('obs-store')
+const punycode = require('punycode')
+const Ens = require('./ens')
+
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+const ZERO_X_ERROR_ADDRESS = '0x'
+
+class EnsController {
+  constructor ({ ens, provider, networkStore } = {}) {
+    const initState = {
+      ensResolutionsByAddress: {},
+    }
+
+    this._ens = ens
+    if (!this._ens) {
+      const network = networkStore.getState()
+      if (Ens.getNetworkEnsSupport(network)) {
+        this._ens = new Ens({
+          network,
+          provider,
+        })
+      }
+
+      networkStore.subscribe((network) => {
+        this.store.putState(initState)
+        this._ens = new Ens({
+          network,
+          provider,
+        })
+      })
+    }
+
+    this.store = new ObservableStore(initState)
+  }
+
+  reverseResolveAddress (address) {
+    return this._reverseResolveAddress(ethUtil.toChecksumAddress(address))
+  }
+
+  async _reverseResolveAddress (address) {
+    if (!this._ens) {
+      return undefined
+    }
+
+    const state = this.store.getState()
+    if (state.ensResolutionsByAddress[address]) {
+      return state.ensResolutionsByAddress[address]
+    }
+
+    const domain = await this._ens.reverse(address)
+    const registeredAddress = await this._ens.lookup(domain)
+    if (registeredAddress === ZERO_ADDRESS) throw new Error('No address for name')
+    if (registeredAddress === ZERO_X_ERROR_ADDRESS) throw new Error('ENS Registry error')
+
+    if (ethUtil.toChecksumAddress(registeredAddress) === address) {
+      this._updateResolutionsByAddress(address, punycode.toASCII(domain))
+      return domain
+    } else {
```

Comment:
should an error be thrown, logged or communicated in this case?

Reply:
It does at least seem inconsistent with the two errors thrown above.

Reply:
The two errors above are handling actual _errors_ similar to the ethjs-ens lib:

https://github.com/ethjs/ethjs-ens/blob/8ea29591ae545a5da243b0f071b5676ff95aa647/index.js#L111-L115

Reply:
Right, but then why isn't it an error if the reverse lookup can't be verified with a forward lookup?

Naively I would think that this function should either 1) return `undefined` when it can't resolve the address, or 2) throw an error when it can't resolve an address. Right now it's throwing an error if it can't reverse resolve it, but returning `undefined` if it can resolve it but can't verify it.

Reply:
Updated